### PR TITLE
Resize watchCache according to watchCacheEvents' density

### PR DESCRIPTION
**Pull request by goodluckbot**
#### What type of PR is this?
/kind feature
/kind cleanup

#### What this PR does / why we need it:
Resize watchCache according to the amount of traffic in the cluster instead of by 2x
We are seeing lots of "too old resource version" error when clients talk to apiserver in our cluster with heavy traffic load.

Resizing watchCache more aggressively has the following benefits 
- save cpu cycle during the resizing procedure under heavy load. e.g. enlarge the cache size from 100 to 800 in just one round instead of three. (100->200->400->800)
- lead to less "too old resource version" error when the churn rate in the cluster is high.

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
Resize watchCache according to the amount of traffic in the cluster instead of by 2x
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

### CODE REVIEW

**Comment by fedebongio** [comment]
Thank you @wojtek-t for taking a look!
/cc @yliaog 
/triage accepted

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 1]
```diff
@@ -343,18 +343,24 @@ func (w *watchCache) updateCache(event *watchCacheEvent) {
...
+// - increases capacity by [ eventFreshDuration / (eventTime - 1st event's timestamp) ]
+// 	 if cache is full and all cached events occurred within last eventFreshDuration.
+// - decreases capacity by [ eventFreshDuration / (eventTime - recent quarter event's timestamp) ] / 2
+//   when recent quarter of events occurred outside of eventFreshDuration(protect watchCache from flapping).
 func (w *watchCache) resizeCacheLocked(eventTime time.Time) {
 	if w.isCacheFullLocked() && eventTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime) < eventFreshDuration {
-		capacity := min(w.capacity*2, w.upperBoundCapacity)
+		timeDiff := float64(eventFreshDuration) / float64(eventTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime))
+		increase := max(int(timeDiff), 2)
+		capacity := min(w.capacity*increase, w.upperBoundCapacity)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:354]: I'm sorry - I lost track of this PR.

I think this is a reasonable improvement. But this change requires unit tests to be mergable.

**Comment by dims** [comment]
@goodluckbot please see last comments ... @wojtek-t would like a unit test to make sure this change keeps working. so can you please add one?

**Pull request closed without merging**